src/infi/clickhouse_orm/query.py

```diff
@@ -342,25 +403,40 @@ def only(self, *field_names):
         qs._fields = field_names
         return qs
 
-    def filter(self, *q, **filter_fields):
+    def _filter_or_exclude(self, *q, **kwargs):
+        reverse = kwargs.pop('reverse', False)
+        prewhere = kwargs.pop('prewhere', False)
+        condition = copy(self._where_q)
+        qs = copy(self)
+
+        for q_obj in q:
+            condition &= q_obj
+
+        if kwargs:
+            condition &= Q(**kwargs)
+
```

Comment:
In both filter and exclude cases, you are ANDing the conditions with the existing ones. I suspect this might change the current behavior (which matches Django's): multiple calls to `filter` are ANDed, but multiple calls to `exclude` are ORed. Consider the two tests I just pushed to the develop branch - do they still pass with the new implementation?

    def test_mutiple_filter(self):
        qs = Person.objects_in(self.database)
        # Single filter call with multiple conditions is ANDed
        self._test_qs(qs.filter(first_name='Ciaran', last_name='Carver'), 1)
        # Separate filter calls are also ANDed
        self._test_qs(qs.filter(first_name='Ciaran').filter(last_name='Carver'), 1)
        self._test_qs(qs.filter(birthday='1970-12-02').filter(birthday='1986-01-07'), 0)

    def test_multiple_exclude(self):
        qs = Person.objects_in(self.database)
        # Single exclude call with multiple conditions is ANDed
        self._test_qs(qs.exclude(first_name='Ciaran', last_name='Carver'), 99)
        # Separate exclude calls are ORed
        self._test_qs(qs.exclude(first_name='Ciaran').exclude(last_name='Carver'), 98)
        self._test_qs(qs.exclude(birthday='1970-12-02').exclude(birthday='1986-01-07'), 98)

Reply:
Hmm, yes it seems I've done a little bit wrong logic.
It would be correct to form current condition at first (via AND), than negate it if needed and then AND it to the original one. I'll fix that on monday.